File: frontend/src/components/game/Navigation/greatCircle.ts
/**
 * Spherical / WGS84-ish navigation mathematics.
 * Great-circle: shortest path on a sphere (orthodrome).
 * Ref: Aviation Formulary (Ed Williams); WGS84 for ECEF helpers.
 */

import { DEG2RAD, EARTH_RADIUS_M, RAD2DEG, WGS84_A, WGS84_E2 } from "../Physics/Math/constants";

const R_NM = EARTH_RADIUS_M / 1852;

export function haversineNm(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  const p1 = lat1 * DEG2RAD;
  const p2 = lat2 * DEG2RAD;
  const dLat = (lat2 - lat1) * DEG2RAD;
  const dLng = (lng2 - lng1) * DEG2RAD;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(p1) * Math.cos(p2) * Math.sin(dLng / 2) ** 2;
  return 2 * R_NM * Math.asin(Math.min(1, Math.sqrt(a)));
}

export function haversineM(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  return haversineNm(lat1, lng1, lat2, lng2) * 1852;
}

/** Initial true bearing (deg) from point 1 → 2. */
export function initialBearingDeg(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  const p1 = lat1 * DEG2RAD;
  const p2 = lat2 * DEG2RAD;
  const dLng = (lng2 - lng1) * DEG2RAD;
  const y = Math.sin(dLng) * Math.cos(p2);
  const x =
    Math.cos(p1) * Math.sin(p2) -
    Math.sin(p1) * Math.cos(p2) * Math.cos(dLng);
  return ((Math.atan2(y, x) * RAD2DEG) + 360) % 360;
}

export function etaSeconds(distanceNm: number, groundSpeedMs: number): number | null {
  if (groundSpeedMs < 5) return null;
  const kt = groundSpeedMs * 1.94384;
  return (distanceNm / kt) * 3600;
}

/**
 * Spherical linear interpolation along the great circle (SLERP on unit sphere).
 */
export function greatCircleWaypoints(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
  steps = 24,
): Array<{ lat: number; lng: number }> {
  const φ1 = lat1 * DEG2RAD;
  const λ1 = lng1 * DEG2RAD;
  const φ2 = lat2 * DEG2RAD;
  const λ2 = lng2 * DEG2RAD;

  const x1 = Math.cos(φ1) * Math.cos(λ1);
  const y1 = Math.cos(φ1) * Math.sin(λ1);
  const z1 = Math.sin(φ1);
  const x2 = Math.cos(φ2) * Math.cos(λ2);
  const y2 = Math.cos(φ2) * Math.sin(λ2);
  const z2 = Math.sin(φ2);

  let d = x1 * x2 + y1 * y2 + z1 * z2;
  d = Math.min(1, Math.max(-1, d));
  const ω = Math.acos(d);
  const out: Array<{ lat: number; lng: number }> = [];

  if (ω < 1e-8) {
    for (let i = 0; i <= steps; i++) out.push({ lat: lat1, lng: lng1 });
    return out;
  }

  const sinω = Math.sin(ω);
  for (let i = 0; i <= steps; i++) {
    const f = i / steps;
    const a = Math.sin((1 - f) * ω) / sinω;
    const b = Math.sin(f * ω) / sinω;
    const x = a * x1 + b * x2;
    const y = a * y1 + b * y2;
    const z = a * z1 + b * z2;
    const φ = Math.atan2(z, Math.hypot(x, y));
    const λ = Math.atan2(y, x);
    out.push({ lat: φ * RAD2DEG, lng: λ * RAD2DEG });
  }
  return out;
}

/** Cross-track error (m) — positive right of course 1→2. */
export function crossTrackErrorM(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
  lat: number,
  lng: number,
): number {
  const d13 = haversineM(lat1, lng1, lat, lng) / EARTH_RADIUS_M;
  const θ13 = initialBearingDeg(lat1, lng1, lat, lng) * DEG2RAD;
  const θ12 = initialBearingDeg(lat1, lng1, lat2, lng2) * DEG2RAD;
  return Math.asin(Math.sin(d13) * Math.sin(θ13 - θ12)) * EARTH_RADIUS_M;
}

/** Geodetic → ECEF (WGS84). */
export function geodeticToEcef(
  latDeg: number,
  lngDeg: number,
  altM: number,
): [number, number, number] {
  const φ = latDeg * DEG2RAD;
  const λ = lngDeg * DEG2RAD;
  const sinφ = Math.sin(φ);
  const cosφ = Math.cos(φ);
  const N = WGS84_A / Math.sqrt(1 - WGS84_E2 * sinφ * sinφ);
  const x = (N + altM) * cosφ * Math.cos(λ);
  const y = (N + altM) * cosφ * Math.sin(λ);
  const z = (N * (1 - WGS84_E2) + altM) * sinφ;
  return [x, y, z];
}

/** Horizon distance (m) for observer height h above sphere. */
export function horizonDistanceM(eyeHeightM: number): number {
  return Math.sqrt(Math.max(0, eyeHeightM * (2 * EARTH_RADIUS_M + eyeHeightM)));
}

File: backend/internal/navigation/greatcircle.go
package navigation

import (
	"math"

	"github.com/mrbishalbaniya/3dearth/backend/internal/domain"
)

const earthRadiusNm = 3440.065

type RouteResult struct {
	DistanceNm float64           `json:"distanceNm"`
	BearingDeg float64           `json:"bearingDeg"`
	ETESec     *int              `json:"eteSec,omitempty"`
	Waypoints  []domain.GeoPoint `json:"waypoints"`
}

func HaversineNm(a, b domain.GeoPoint) float64 {
	p1 := a.Lat * math.Pi / 180
	p2 := b.Lat * math.Pi / 180
	dLat := (b.Lat - a.Lat) * math.Pi / 180
	dLng := (b.Lng - a.Lng) * math.Pi / 180
	h := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(p1)*math.Cos(p2)*math.Sin(dLng/2)*math.Sin(dLng/2)
	return 2 * earthRadiusNm * math.Asin(math.Min(1, math.Sqrt(h)))
}

func InitialBearingDeg(a, b domain.GeoPoint) float64 {
	p1 := a.Lat * math.Pi / 180
	p2 := b.Lat * math.Pi / 180
	dLng := (b.Lng - a.Lng) * math.Pi / 180
	y := math.Sin(dLng) * math.Cos(p2)
	x := math.Cos(p1)*math.Sin(p2) - math.Sin(p1)*math.Cos(p2)*math.Cos(dLng)
	brg := math.Atan2(y, x) * 180 / math.Pi
	return math.Mod(brg+360, 360)
}

func GreatCircle(a, b domain.GeoPoint, steps int, cruiseSpeedMs float64) RouteResult {
	if steps < 2 {
		steps = 16
	}
	// Spherical SLERP via ECEF
	φ1, λ1 := a.Lat*math.Pi/180, a.Lng*math.Pi/180
	φ2, λ2 := b.Lat*math.Pi/180, b.Lng*math.Pi/180
	x1, y1, z1 := math.Cos(φ1)*math.Cos(λ1), math.Cos(φ1)*math.Sin(λ1), math.Sin(φ1)
	x2, y2, z2 := math.Cos(φ2)*math.Cos(λ2), math.Cos(φ2)*math.Sin(λ2), math.Sin(φ2)
	d := math.Acos(math.Max(-1, math.Min(1, x1*x2+y1*y2+z1*z2)))
	wps := make([]domain.GeoPoint, 0, steps+1)
	if d < 1e-9 {
		for i := 0; i <= steps; i++ {
			wps = append(wps, a)
		}
	} else {
		sinD := math.Sin(d)
		for i := 0; i <= steps; i++ {
			f := float64(i) / float64(steps)
			A := math.Sin((1-f)*d) / sinD
			B := math.Sin(f*d) / sinD
			x, y, z := A*x1+B*x2, A*y1+B*y2, A*z1+B*z2
			φ := math.Atan2(z, math.Hypot(x, y))
			λ := math.Atan2(y, x)
			wps = append(wps, domain.GeoPoint{Lat: φ * 180 / math.Pi, Lng: λ * 180 / math.Pi})
		}
	}
	dist := HaversineNm(a, b)
	res := RouteResult{
		DistanceNm: dist,
		BearingDeg: InitialBearingDeg(a, b),
		Waypoints:  wps,
	}
	if cruiseSpeedMs > 1 {
		kt := cruiseSpeedMs * 1.94384
		sec := int((dist / kt) * 3600)
		res.ETESec = &sec
	}
	return res
}

func FuelEstimateKg(distanceNm, burnKgPerHour, reserveKg float64) float64 {
	hours := distanceNm / 120 // fallback 120 kt average if unknown
	if hours < 0.1 {
		hours = 0.1
	}
	return burnKgPerHour*hours*0.85 + reserveKg
}
